Hometask7

MihailNenahov opened the pull request:
I still have some questions, so please look at commit history and write some comments about what was wrong.

Commit:
not working version

i have question, why it doesn't works

Commit:
not working

another version, but i still don't understand, why it doesn't works

Commit:
hometask 7

first task sends json format

second task implements tcp server-client connedcion(i left some questions in first two commits)

third task sends form and create cookie

Commit:
hometask 7

first task sends json format

second task implements tcp server-client connedcion(i left some questions in first two commits)

third task sends form and create cookie

imatsiash-ib commented on first_task/hello_world/hello_world.go:
```diff
@@ -1,9 +0,0 @@
-//i'll left some questions
```
Why I can see redundant files in this PR ?

imatsiash-ib commented on task1/main.go:
```diff
@@ -0,0 +1,34 @@
+	Host       string      `json:"host"`
+	UserAgent  string      `json:"user_agent"`
+	RequestUri string      `json:"request_uri"`
+	H          http.Header `json:"headers"`
```
H ? What does it mean ?
Please use more obvious names.

imatsiash-ib commented on task1/main.go:
```diff
@@ -0,0 +1,34 @@
+	}
+
+	w.Header().Set("Content-Type", "applicatioin/json; charset=UTF-8")
+	buf, _ := json.Marshal(ans)
```
Why do you ignore an error returned by `Marshal` function ?

Commit:
after reading comments

removed redundant files
renamed H to Header
added err check for Marshal func

added err check in error case

imatsiash-ib commented on task1/main.go:
```diff
@@ -0,0 +1,38 @@
+
+	buf, err := json.Marshal(ans)
+	if err != nil {
+		log.Println(err)
```
In case of error you should return error message to the client, and set status code to an error.

wshaman commented on task1/main.go:
```diff
@@ -0,0 +1,39 @@
+
+	buf, err := json.Marshal(ans)
+	if err != nil {
+		fmt.Fprint(w, err)
```
Error header should be set here.

imatsiash-ib replied:
Yes, you should return error to the user.

wshaman commented on task2/server/server.go:
```diff
@@ -0,0 +1,62 @@
+		conn, err := ln.Accept()
+		fmt.Println(conn.RemoteAddr().String(), " connected")
+		if err != nil {
+			fmt.Print("2) ")
```
`log.Panic` does the same.

wshaman commented on task3/error.html:
```diff
@@ -0,0 +1,11 @@
+
+<h2>404</h2>
+
+<p>my teacher told me to not show sense of error to user, so u just need to know, that smth went wrong</p>
```
custom 404 is a cool tradition.

wshaman commented on task3/index.html:
```diff
@@ -0,0 +1,21 @@
+<!DOCTYPE html>
+<html>
```
any place to show cookie?

wshaman commented on task3/main.go:
```diff
@@ -0,0 +1,66 @@
+		w.WriteHeader(http.StatusMethodNotAllowed)
+		t, _ := template.ParseFiles("error.html")
+		t.Execute(w, errorToSend{
+			http.StatusText(http.StatusMethodNotAllowed),
```
I got the idea, anyway here should be different mesages

wshaman reviewed (approved):
That's OK. Only one thing isn't working for me, it's an HTML layout in 3rd task, but it's not critical, we are not learning HTML here.

imatsiash-ib commented on task2/client/client.go:
```diff
@@ -0,0 +1,48 @@
+
+		data, err := input.ReadString('\n')
+
+		if err != nil {
```
Please remove all extra empty lines.

imatsiash-ib replied:
Please do not ignore an error.

imatsiash-ib commented on task2/client/client.go:
```diff
@@ -0,0 +1,48 @@
+
+		check := strings.TrimSuffix(data, string('\n'))
+
+		if check == "EXIT" {
```
What happens if I write `exit` ?

MihailNenahov replied:
`exit` is not a command for exit. if `exit` is input, so this string will be sent to server and server will process this string and will return `EXIT`.

imatsiash-ib replied:
From homework description:
C exits by input `exit`.

I expect that when client input == `exit`, client should be stopped. Am I correct?

MihailNenahov replied:
u are right. my fault.

imatsiash-ib commented on task2/server/server.go:
```diff
@@ -0,0 +1,62 @@
+	ln, err := net.Listen("tcp", ":8083")
+	defer ln.Close()
+	if err != nil {
+		log.Fatal("1) ", err)
```
What `1) ` does mean ?

MihailNenahov replied:
it helps to find a place, where program felt.

imatsiash-ib replied:
This is a debug information, that will be redundant for the client.

imatsiash-ib commented on task2/server/server.go:
```diff
@@ -0,0 +1,62 @@
+			if err != io.EOF {
+				fmt.Println(err)
+			}
+			break
```
Strange construction..

Commit:
after reading comments

removed empty lines
in task2/server rewrote error handle
added in task1 header content
in task3 wrote different messages

Commit:
after reading comments

changed exit condition
removed debuging info
